frame/support/src/evm.rs

```diff
@@ -0,0 +1,109 @@
+// This file is part of Darwinia.
+//
+// Copyright (C) 2018-2021 Darwinia Network
+// SPDX-License-Identifier: GPL-3.0
+//
+// Darwinia is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Darwinia is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Darwinia. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+// --- crates.io ---
+use ethereum::TransactionMessage;
+use sha3::{Digest, Keccak256};
+// --- darwinia-network ---
+use ethereum_primitives::{H160, H256, U256};
+// --- paritytech ---
+use frame_support::PalletId;
+use sp_std::vec;
+use sp_std::vec::Vec;
+
+pub const POW_9: u32 = 1_000_000_000;
+pub const INTERNAL_TX_GAS_LIMIT: u32 = 300_000_000;
+pub const SELECTOR: usize = 4;
+pub const TRANSFER_ADDR: &'static str = "0x0000000000000000000000000000000000000015";
+
+pub trait IntoDvmAddress {
+	fn into_dvm_address(&self) -> H160;
+}
+
+// Convert pallet id to dvm address
+impl IntoDvmAddress for PalletId {
+	fn into_dvm_address(&self) -> H160 {
+		let mut bytes = vec![0u8; 12];
+		bytes.append(&mut self.0.to_vec());
+		H160::from_slice(&bytes)
```

Comment:
What's the H160 looks like?

It looks it will have prefix with 12 0u8, [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, self's elements ] , which is wired.

Document the spec here, and refer the into_sub_account of PalletId:

```
/// Format is TYPE_ID ++ encode(parachain ID) ++ 00.... where 00... is indefinite trailing zeroes to
/// fill AccountId.
impl<T: Encode + Decode + Default, Id: Encode + Decode + TypeId> AccountIdConversion<T> for Id {
	fn into_sub_account<S: Encode>(&self, sub: S) -> T {
		(Id::TYPE_ID, self, sub)
			.using_encoded(|b| T::decode(&mut TrailingZeroInput(b)))
			.unwrap_or_default()
	}

	fn try_from_sub_account<S: Decode>(x: &T) -> Option<(Self, S)> {
		x.using_encoded(|d| {
			if &d[0..4] != Id::TYPE_ID {
				return None
			}
			let mut cursor = &d[4..];
			let result = Decode::decode(&mut cursor).ok()?;
			if cursor.iter().all(|x| *x == 0) {
				Some(result)
			} else {
				None
			}
		})
	}
}
```

Reply:
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.append

Reply:
I change the `into_dvm_address` impl. Please check again.